From: SJ Park <sj@kernel.org>
Date: Tue, 4 Aug 2026 07:37:00 -0700
Subject: [PATCH] mm/damon/core: document damon_call()/damon_start() race hang
 issue

Let's suppose damon_start() an damon_call() are executed in parallel for
same DAMON context.  Then, damon_call() could show
ctx->damon_calls_obsolte set while ctx->kdamond is unset.  If
damon_start() sets ctx->kdamond before damon_call() startis the
cancelling, damon_call() can indefinitely hang.  No DAMON API caller do
such parallel execution of damon_start() and damon_call(), so the issue
doesn't exist.  Add a clarification comment for caution.

Signed-off-by: SJ Park <sj@kernel.org>
---
 mm/damon/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index cf10f6d725c19..2817aab7de276 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2251,6 +2251,9 @@ int damon_kdamond_pid(struct damon_ctx *ctx)
  *
  * When this function is failed, the @ctx is guaranteed to be stopped.
  *
+ * This function should not be called in parallel to damon_start() for the
+ * @ctx.  In the case, this function could indefinitely hang.
+ *
  * Return: 0 on success, negative error code otherwise.
  */
 int damon_call(struct damon_ctx *ctx, struct damon_call_control *control)
-- 
2.47.3

